src/ctia/task/migration/migrate_es_stores.clj

```diff
@@ -75,6 +75,79 @@
+  `data-chan`, along with next search_after value for eventual further restart. Once all the
+   documents that match given query are retrieved, it closes the channel `data-chan`"
+  [{:keys [source-store
+           search_after
+           data-chan
+           batch-size
+           query]}]
+  (loop [current-search-after search_after]
+    ;; loop that reads batches from source and produces them in chan
+    (let [{:keys [data paging] :as res} (mst/query-fetch-batch query
+                                                               source-store
+                                                               batch-size
+                                                               0
+                                                               "asc"
+                                                               current-search-after)
+          next (:next paging)
+          next-search-after (:sort paging)]
+      (>!! data-chan {:documents data
+                      :search_after next-search-after})
+      (if next
+        (recur  next-search-after)
+        (close! data-chan)))))
+
+(defn write-target
+  "This function reads batches from `data-chan` (pushed by read-source) until it reads
+   a `nil`, meaning that the channel has been closed by `read-source` function. The
+   batches are (1) modified with `migrations` functions, (2) validated by `list-coerce`
+   and (3) written into given `target-store`. At each batch the migration state, number
+   of migrated documents and search_after value, is updated in ES after the documents
+   are written in the store. If the process fails or is stopped before that update of
+   the migration state, the documents will be red again by `read-source` and overridden
+   in case of restart."
+  [{:keys [target-store
+           data-chan
+           migrations
+           entity-type
+           batch-size
+           migration-id
+           list-coerce
+           migrated-count
+           migration-es-conn
+           confirm?]}]
+  (loop [current-count migrated-count]
+    ;; sync loop that consumes batches in chan and migrates them to target
+    (if-let [{:keys [documents search_after]} (<!! data-chan)]
+      (let [migrated (transduce migrations conj documents)
+           {:keys [data errors]} (list-coerce migrated)
+           new-current-count (+ current-count (count data))]
+        (doseq [entity errors]
+          (let [message
+                (format "%s - Cannot migrate entity: %s"
+                        (name entity-type)
+                        (pr-str entity))]
+            (log/error message)))
+        (when confirm?
+          (when (seq data) (mst/store-batch target-store data))
+          (mst/rollover target-store batch-size new-current-count)
+          (mst/update-migration-store migration-id
+                                      entity-type
+                                      (into {:target {:migrated new-current-count}}
```

Comment:
it would look better if you use `cond->` start with `{:target {:migrated new-current-count}}` and assoc only if `search_after` is non nil